ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,575 @@
+    class. Basically at the top of these calls there's a `if self != Workflow` check.
+* `respond_to_missing?` + `inherited` will be leveraged such that all definition-time defined handlers will return their
+  definition if they are called as class methods. So `FooWorkflow.some_signal` will return a
+  `Workflow::Definition::Signal` if an instance method of `some_signal` exists marked as a `workflow_signal`.
+  * 💭 Why? It is helpful for callers/client-side to be able to reference something more explicit than a symbol, and it
+    also helps with making sure if the name is changed we use that name instead of the method name. Symbols can always
+    be used though.
+  * ❓ Is this acceptable or too much magic?
+  * ❓ Is there a better way to reference an interaction outside of the instance of the class that works well with Ruby
+    generic typing frameworks, e.g. Sorbet?
+    * We are trying to avoid the proxy model that Java takes because we have found it is too confusing to users to hide
+      that method invocation is actually, say, activity invocation.
+
+## Dynamic Workflows, Signals, Queries, and Updates
+
+* Support dynamic pieces similar to how we have done in Python and .NET.
+* Create a new `Temporalio::Converters::RawValue` object that is a pass-through object for raw values (still subject to
+  payload codecs), and a `payload_converter` is available in workflow/activity context to convert it.
+  * Handler methods will use `RawValue` for args if told to, similar with workflow `execute`/`initialize` if
+    `workflow_raw_args` is set.
+  * `RawValue` can be returned too.
+* To mark a workflow as dynamic, `workflow_dynamic` class method is called. Only one dynamic workflow can be present for
+  a worker. The workflow can retrieve its workflow type name via `Workflow.info`
+* Dynamic signals, queries, and updates are enabled via `dynamic: true` on the calls before the methods.
+  * Dynamic handlers first argument is the string name (it's not a symbol), and all successive arguments are the actual
+    arguments.
+  * Only one dynamic handler of a type (signal, query, or update) can be on a workflow.
+
+## Workflow Asynchrony
+
+Ruby does not have a built-in task/promise/future/etc concept. However it does have fibers and we can leverage those.
+
+### Custom Fiber Scheduler
+
+We will write a custom Fiber scheduler. This will provide the following benefits:
+
+* Can leverage `Timeout` for timeouts.
+* Can leverage `sleep` for sleeps.
+* Able to fail some IO calls.
+* Allows usage of most of https://github.com/socketry/async.
+  * We will have many tests confirming many behaviors here but we still get to avoid a runtime dependency.
+  * ❓ Why not just take a dependency on `async` gem? It does have a deterministic scheduler and does a decent job of
+    everything we want, but not only do we want to avoid transitive dependencies in our library (for versioning and
+    other reasons), we cannot rely on its logic into the future.
```

Comment:
Is `async`s deterministic behavior actually part of its contract or no? (I'm thinking of how we got burned by Python's change in behavior for asyncio; I imagine you are too)

Reply:
> Is asyncs deterministic behavior actually part of its contract or no?

Not explicitly, but most of the things it does are delegated to the underlying fiber scheduler. The other problem is not just that it is deterministic, but that it also doesn't make changes that would be incompatible (e.g. schedule two fibers in a different order).

Concerning determinism, they do say as a comment inside https://github.com/socketry/async/blob/72fae62658b68e85d53d4c6adf13d3b34108cfe8/lib/async/scheduler.rb#L428-L435 that they intentionally try to preserve determinism, but obviously a code comment is not a clear contract (and that's their scheduler, we'll have our own).

It's a tough call, because it's a full featured library with lots of utilities that leans on the deterministic fiber scheduler, and we surely don't want to recreate _every_ construct they have already created. But maybe we don't want to encourage use of this project? Java SDK gets along just fine with its primitive `Promise` support. We still want to lean on traditional Ruby things like queues, timeouts, and sleep I think, but we'd have to write our own mutexes, semaphores, etc.

Reply:
Yeah, this would be a much easier call if determinism were an explicit design goal, but since it's not as such, I think it's probably safer to avoid the dependency for now. Perhaps we could align our own API surface with its API so we could switch later if there is user demand? That might also help reduce the cognitive overhead of learning Temporal if the user already knows `async`.

Reply:
> I think it's probably safer to avoid the dependency for now

To clarify, we would never have a dependency, just that it would happen to work and we could recommend its use (and we'd test that it does in our tests). Having said that, I think I agree maybe we should not even recommend its use or discuss it.

> Perhaps we could align our own API surface with its API so we could switch later if there is user demand

It's too large. I think we can have `Future` (what they call `Task`) and `Mutex` manually implemented, and then decide whether to explicitly implement `sleep` and `Timeout` and `Queue` or allow reuse of the standard library ones (knowing in Ruby this isn't the same as leaning on a large set of async primitives, these are 3 very limited low-level common Ruby things).

I think I'm leaning towards no `async` library recommendation, `Future` + `Mutex` implemented, `sleep` + `Timeout` + `Queue` just work from the standard library.

Reply:
Proposal updated to discourage use of `async` library and clarify that we do support stdlib things that work in all environments.